### PR TITLE
REST API for looking up metadata by IRI

### PR DESCRIPTION
**Pull request by wgwz**
## Description

Closes: https://github.com/regen-network/regen-web/issues/841

This adds a REST endpoint for looking up a JSON-LD metadata by IRI. Thanks @blushi for the pair programming on this!

Currently the endpoint works as follows:

1. For a non-existent IRI, we are returning a 404 error with a descriptive message:

```
kyle@lookfar4:~$ http :5000/metadata-graph/regen:SeKXt1q9UEgBDEcAJF9hkadSFfFGnrCKyeqSyZ74G31s4tmz.rdf
HTTP/1.1 404 Not Found
Connection: keep-alive
Content-Length: 96
Content-Type: text/html; charset=utf-8
Date: Mon, 14 Feb 2022 18:10:06 GMT
ETag: W/"60-u6N0zunA75oUbk8a9VC284gYmYg"
Keep-Alive: timeout=5
Vary: Origin
X-Powered-By: Express

metadata_graph with the iri regen:SeKXt1q9UEgBDEcAJF9hkadSFfFGnrCKyeqSyZ74G31s4tmz.rdf not found
```

2. For a existing IRI, we return the metadata as an object:

```
kyle@lookfar4:~$ http :5000/metadata-graph/regen:SeKXt1q9UEgBDEcAJF9hkadSFfFGnrCKyeqSyZ74G31s4tmz.rdf
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 26
Content-Type: application/json; charset=utf-8
Date: Mon, 14 Feb 2022 18:09:06 GMT
ETag: W/"1a-PLGdsWkqJF+sQ7TUabYY6UXz3Pc"
Keep-Alive: timeout=5
Vary: Origin
X-Powered-By: Express

{
    "metadata": {
        "foo": "bar"
    }
}
```

3. For an invalid IRI, we return a 404 error with a descriptive message:

```
kyle@lookfar4:~$ http :5000/metadata-graph/test
HTTP/1.1 400 Bad Request
Connection: keep-alive
Content-Length: 56
Content-Type: text/html; charset=utf-8
Date: Mon, 14 Feb 2022 18:06:00 GMT
ETag: W/"38-HO2/efAsKjYB4BIqNE3wytSKZ1o"
Keep-Alive: timeout=5
Vary: Origin
X-Powered-By: Express

Invalid IRI, it must be of the form regen:<iri-hash>.rdf
```
---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] targeted `dev` branch
- [x] provided a link to the relevant issue or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all tests passed
- [ ] submit a PR against `master` branch after this one (and other PRs depending on this one, e.g. on regen-network/regen-web, if applicable) get merged

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

File: server/routes/metadata-graph.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+const router = express.Router();
+
+router.get('/metadata-graph/:iri', async (req, res) => { 
+  const { iri } = req.params;
+  const iri_re = new RegExp("regen:.+\.rdf");
+  let client;
+  try {
+    if (!iri_re.test(iri)) {
+      res.status(400).send("Invalid IRI, it must of the form regen:<iri>.rdf");
```
[the remark this answers is not in the document]
Comment (blushi) [reply server/routes/metadata-graph.ts:13]: I think that makes sense so we avoid any database query if we know the iri is invalid anyway!

File: server/routes/metadata-graph.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    } else {
+      client = await pgPool.connect();
+      try {
+         const { rows } = await client.query(
+           'select metadata from metadata_graph where iri=$1',
+           [iri]
+         )
+         res.json(rows);
```
[the remark this answers is not in the document]
Comment (blushi) [reply server/routes/metadata-graph.ts:21]: I believe we should rather check for the length of `rows`, if 0 then return a 404 with some json error message `metadata_graph with the iri ${iriValue} not found`, else return `rows[0]` with 200 code

**Review by blushi — COMMENTED** [review]
Looks good overall, just a small comment

File: server/routes/metadata-graph.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+const router = express.Router();
+
+router.get('/metadata-graph/:iri', async (req, res) => { 
+  const { iri } = req.params;
+  const iri_re = new RegExp("regen:.+\.rdf");
+  let client;
+  try {
+    if (!iri_re.test(iri)) {
+      res.status(400).send("Invalid IRI, it must of the form regen:<iri-hash>.rdf");
```
Comment (blushi) [inline server/routes/metadata-graph.ts:13]:
```suggestion
      res.status(400).send("Invalid IRI, it must be of the form regen:<iri-hash>.rdf");
```

File: server/routes/metadata-graph.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    } else {
+      client = await pgPool.connect();
+      try {
+         const { rows } = await client.query(
+           'select metadata from metadata_graph where iri=$1',
+           [iri]
+         )
```
Comment (blushi) [inline server/routes/metadata-graph.ts:20]:
```suggestion
         );
```

File: server/routes/metadata-graph.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+         if (!rows.length) {
+           res.status(404).send(`metadata_graph with the iri ${iri} not found`)
```
Comment (blushi) [inline server/routes/metadata-graph.ts:22]: We definitely need to add some prettier/eslint to this repo, created an issue for that: https://github.com/regen-network/regen-registry/issues/830

```suggestion
           res.status(404).send(`metadata_graph with the iri ${iri} not found`);
```

File: server/routes/metadata-graph.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    }
+  } catch (err) {
+    console.error('Error acquiring postgres client', err);
+    res.sendStatus(500);
+  } finally {
+    if (client) {
+      client.release();
+    }
+  }
+})
```
Comment (blushi) [inline server/routes/metadata-graph.ts:39]:
```suggestion
});
```

**Review by blushi — COMMENTED** [review]
Looks good, pending some nits.
Could you update the PR description to reflect your latest changes wrt the response if not found and resolve the conflicts (you might need to rm all your node_modules folder, delete your `server/yarn.lock` and re-run `yarn`)?

File: server/routes/metadata-graph.ts (hunk 5 above)
Comment (wgwz) [reply server/routes/metadata-graph.ts:22]: I'll be glad to pick that one up, it should be a pretty quick task.

File: sql/V1_78__project_issuer.sql [hunk 7]
```diff
@@ -0,0 +1,4 @@
+-- Renaming this to be more consistent with ledger data
```
Comment (haveanicedavid) [inline sql/V1_78__project_issuer.sql:1]: 👍

**Review by haveanicedavid — APPROVED** [review]

**Review by blushi — APPROVED** [review]
LGTM

**Pull request merged**